### ISSUE TITLE
::object.func !== ::object.func

### ISSUE THREAD

**Issue by lukescott** [zenparsing/es-function-bind#17]
I'm concerned that this syntax can be very misleading. For example:
```javascript
var object = {
  clickHandler() {}
}

element.addEventListener("click", ::object.clickHandler);
// later in the code...
element.removeEventListener("click", ::object.clickHandler); // doesn't work!

console.log(::object.clickHandler === ::object.clickHandler); // false
```
It's not immediately clear the two are not the same thing - something new is being created each time, which is being masked by the hidden bind.

**Comment by zenparsing** [comment]
So there's some footgun potential here.  This is a good point - thanks for bringing it up.

**Comment by shovon** [comment]
Which is precisely why [azu/idempotent-bind](https://github.com/azu/idempotent-bind) was created.

Not sure how it can be handled with this new syntax, because all idempotent-bind does is a form of memoization, where as the syntax uses `Function.prototype.bind`.

**Comment by zenparsing** [comment]
@shovon Unfortunately, there's no good place to store the memoization without "infecting" the entire object model.  I see you're using WeakMap (which is cool) but that's not something we'd want to put into the core of the language.

@domenic what do you think?  I feel like this might be fatal for the unary operator.

**Comment by domenic** [comment]
IMO this is totally fine... A few arguments for it:

- In general `<unary operator> <member expression>` is not equal to itself. `delete x.y !== delete x.y`; `++x.y !== ++x.y`; even `!x.y !== !x.y` in general if `y` is a getter. Perhaps the strongest analogy is `new x.y !== new x.y`.
- If you think of operators as function calls, it's completely unsurprising for a function to behave this way.
- If you understand what binding is, it's completely unsurprising for the bind operator to behave this way.
- If you think of `::x.y` as shorthand for `x::x.y`, it'd be very surprising for either: (a) the identity behavior to diverge between them; or (b) for `x::x.y` to have identity when repeated where `z::x.y` does not. (Unless the proposal is to make arbitrary `a::b` memoized?)

Maybe @erights has more thoughts, since I know he was a fan of the unary operator version.

**Comment by erights** [comment]
@shovon , the memoization would be fatal unless you froze the bound function. Otherwise, merely sharing a frozen query-only-function would enable a communications channel through the memoization table. For example

Assume a SES or SES-like environment in which all the primordials are frozen.

Assume at this point that we have access to alice and bob that do not have any means to communicate with each other unless we provide such access. In other words, neglecting transitively immutable objects, the only connectivity between alice and bob at this moment is that we have access to both. Neither alice nor bob have access, direct or indirect, to each other.

const f = Object.freeze(function{return 3;});

alice(f);

bob(f);

Here, we have shared with alice and bob only a transitively immutable object, which should therefore not enable them to communicate. However

alice says:

const bound = ::f.toString;
bound.x = somethingIWantToTellBob;

bob says:

const bound = ::f.toString;
const somethingAliceWantedToTellMe = bound.x;

If there's no memoization, then alice and bob get separate bound functions that share only f, thereby not providing any mutability beyond that provided by f. On the other hand, if ::f.toString meant approximately

Object.freeze(f.toString.bind(f))

then the bound object would still not provide any mutability beyond that provided by f. In that case, memoizing it would not be fatal. But JS programmers might find this freezing surprising.

**Comment by erights** [comment]
@domenic I agree, but your first few examples weaken your argument. Unlike "delete" or "++", there's nothing side-effectly about ::f.x beyond the creation of a fresh identity. Your "new" example is perfect.

**Comment by shovon** [comment]
> I see **you're** using WeakMap

@zenparsing Off-topic. I love the recognition, but I can't take credit for [azu/idempotent-bind](https://github.com/azu/idempotent-bind). I did however write [shovon/idempotent.js](https://github.com/shovon/idempotent.js), an unrelated library.

I know, it's easy to conflate between the two.

**Comment by shovon** [comment]
@erights that wasn't my point. I am fully aware of the plethora of things that *can* go wrong with sharing references, including sharing references to a function.

My point was that the "secret sauce" behind being able to generate a bound equivalent of a function in an idempotent manner was to use memoization. No where did I mention it was a silver bullet.

**Comment by lukescott** [comment]
@domenic When someone first looks at `::x.y` they may not readily make the comparison to `new x.y`. I believe that the same problem also applies to `x::x.y`. The syntax isn't evident that something new is being created each time. The `::` operator is also used in several other languages as a _static_ operator. Using it in a dynamic context, even for existing JavaScript developers, will be quite surprising.

For example:
```javascript
function foo() {
}
var foo1 = foo;
var foo2 = foo;
console.log(foo1 === foo2); // true
```
vs
```javascript
function foo() {
}
var obj = {};
var foo1 = obj::foo;
var foo2 = obj::foo;
console.log(foo1 === foo2); // false?
```
vs
```javascript
function foo() {
}
var obj = {};
var foo1 = foo.bind(obj);
var foo2 = foo.bind(obj);
console.log(foo1 === foo2); // definitely false!
```

Memorization in some form would fix this. Browser vendors would likely implement this in a different way than made possible in a pollyfill. For example, modifying `::x.y` may end up modifying `y`. Since this isn't possible in a pollyfil, freezing the memorized function would be prudent to prevent such a pattern from being relied upon.

Given this is an ES7 proposed syntax, adding a static identifier with Object.defineProperty (w/ all the options set to false) to support the memorization wouldn't necessarily be a bad thing. It would only really be an issue with IE 8 and below - depending on the implementation (you would have more of an issue if adding an identifier to `x` of `x::y` than `y` since you are most likely to enumerate over `x`).

**Comment by zenparsing** [comment]
My feeling right now on this issue:

- The "non-equality" of extracted methods will be surprising to some users.  In other languages (e.g. Python, Ruby), extracted methods have equality.
- Introducing a global memozation table of frozen bound methods to resolve the equality problem seems *really* heavy-handed, and I think it will bust our complexity budget for this feature.
- The "frozen"-ness of bound methods coming out of this table will likely be surprising to users anyway (although perhaps less so than inequality).

**Comment by erights** [comment]
You didn't state a conclusion though. Reading only slightly between the
lines, you recommend that we pay the surprise-cost of inequality? That
sounds fine to me.



On Sat, Jun 6, 2015 at 10:38 AM, zenparsing <notifications@github.com>
wrote:

> My feeling right now on this issue:
>
>    - The "non-equality" of extracted methods will be surprising to some
>    users. In other languages (e.g. Python, Ruby), extracted methods have
>    equality.
>    - Introducing a global memozation table of frozen bound methods to
>    resolve the equality problem seems *really* heavy-handed, and I think
>    it will bust our complexity budget for this feature.
>    - The "frozen"-ness of bound methods coming out of this table will
>    likely be surprising to users anyway (although perhaps less so than
>    inequality).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zenparsing/es-function-bind/issues/17#issuecomment-109622155>
> .
>



-- 
Text by me above is hereby placed in the public domain

  Cheers,
  --MarkM

**Comment by zenparsing** [comment]
Very astute @erights  : )

I was avoiding stating a conclusion because I'm not sure what I want yet, and I'm not sure it will be well-received.

First, I don't think we can get away with a memoization table.

But stepping back, what I *really* want is extension methods.  I think this thing is awesome:

https://github.com/gsklee/bound-native-methods

And personally, I would be fine dropping method extraction to get just extension methods.  Others might feel differently though, and that's OK.

**Comment by erights** [comment]
> And personally, I would be fine dropping method extraction to get just extension methods

What are you proposing? What does "get just extension methods" mean? I don't get it.

**Comment by nathggns** [comment]
To be honest, I think the prefix form `::console.log` is going to be far more welcomed than the `arr::map(..)` form.

**Comment by zenparsing** [comment]
Sorry, I shouldn't introduce terminology into the middle of a discussion without defining it.

I just mean the binary form of `::` in this proposal.

**Comment by Naddiseo** [comment]
@zenparsing wrote:
> Unfortunately, there's no good place to store the memoization without "infecting" the entire object model.

@lukescott wrote:
> Given this is an ES7 proposed syntax, adding a static identifier with Object.defineProperty (w/ all the options set to false) to support the memorization wouldn't necessarily be a bad thing.

Am I correct in thinking you're both referring to something like adding a `[@@boundTo]` symbol/slot to a method? Which, without too much thinking, is the best way I can come up with to solve memoization without a global object.

**Comment by lukescott** [comment]
@Naddiseo I'm not sure memorization is really needed. The two use cases from the examples in the proposal seem to be about solving: function changing for infix (utility libs) and method binding for prefix. Memorized binding is only needed as a result of foot stomping from a bi-product of the actual problems trying to be solved. As discussed in #18, if we split the two and work to solve those problems specifically, memorization should not be needed. -- Assuming of course that my assumptions are correct.

**Comment by zenparsing** [comment]
@erights In your opinion, is a memoization table of frozen bound functions arising out of this operator something doable from an implementation point of view?  Like I said before, it seems really heavy-handed, but I don't want to assume anything.

**Comment by erights** [comment]
If I thought

    ::object.func === ::object.func

was an essential feature, I would implement the semantics a different way. Note that === must already case split before it compares, where it pointer compares only on one of the cases. Thus, having instances of value types compare by contents won't cause === to be slower on normal objects, so neither should the following proposal.

Internally, I would have the kinds of frozen bound function that you're talking about not actually be memoized, but rather, compare as === based on being the same binding of the same underlying function. If we get the freezing right, there should be *no* observable difference between such an implementation and a memoizing implementation. In fact, this very question is a good test of whether we got the freezing right.

Well, except for one thing :(.

We can explain why instances of value types cannot serve as keys in WeakMaps. However, for this internal equality change to not be observably different than memoization, these bound functions would need to be able to be WeakMap keys looked up according to these equality behaviors. In order for the WeakMap to serve its (technically unobservable) purpose, these entries in the WeakMap should still be collected when an equivalent key can no longer be created, i.e., when the underlying function and bound value are also deemed unreachable.

Having now gone through the implementation complexity on this path, I continue to advocate that

    ::object.func !== ::object.func

is the best solution.

**Comment by Naddiseo** [comment]
@erights, or solve it at the origin of the problem: where the method is defined (as @lukescott has suggested)

**Comment by lukescott** [comment]
For reference, I'm suggesting:
```javascript
class Foo {
  handleEvent() => {
      // this is bound to instance
  }
}
```
Which is what #15 , is suggesting, but re-use the `=>`. I think `::object.func` is overkill if we can solve it at the source.

**Comment by domenic** [comment]
"Solving it at the source" is exactly *wrong*. We do not want a language feature to encourage that kind of bad programming. Instead we want one to make *good* programming easier, where you only create per-instance bound versions as necessary.

**Comment by erights** [comment]
Aha! After @Naddiseo 's message I tried to find which of @lukescott 's messages was meant. I completely missed this one and have not previously seen this suggestion. Interesting. I have to think about this.

**Comment by erights** [comment]
A lot is said at #15 and I am not finding this specifically. Can you point me at where this is explained, or better, state a self-contained explanation here? Can this syntax be explained by expansion to the existing language? Even approximately with prose stating remaining differences?

**Comment by lukescott** [comment]
Ah, sorry it was #16. It was mentioned as an additional syntax. There was some back and forth on several topics in #18. I'm suggesting that the idea in #16 replace `::object.func` and re-use the fat-arrow syntax already in ES6.

**Comment by Naddiseo** [comment]
@domenic, forgive my ignorance here, but why is specifying it "bad" programming? 

Are you saying this:
```javascript
// This is bad?
function A() { this.bound = (function(){}).bind(this) };

// Because you're binding every instantiation
let objs = [];
for (i = 0; i <10; i++) {
    objs.push(new A());
}

// Even though you only maybe use the bound method for one instance?
let ret = objs[0].bound();
```

My counter example (which will probably show my ignorance):
```javascript
function A() { this.method = function() {} };

function makeVirtualDom(ctrl) {
   // the method is rebound every time the dom is redrawn.
   return React.createElement('div', { onclick: ctrl.method.bind(ctrl) });
}

// but the constructor is only created once (which is why I was thinking "solving it at the source" would be better.
let ctrl = new A();
// "Redraw loop"
for (i = 0; i<10; i++) {
   redraw(makeVirtualDom(ctrl));
}
```

However, my counter example is only that: one specific usecase. But, the virtual dom callback situation is by far the biggest use of `.bind` in my code bases. @domenic, are you suggesting this pattern can be solved in a different way?

**Comment by lukescott** [comment]
@domenic Allowing methods to be copied from a class is bad programming. Other languages don't allow this. But it's already part of the language. Since it's already too late for that, I think it would be prudent to fix the binding at the source. Most of the time this is needed for event handlers / callbacks. Allowing an extension to `=>` would be a good way of documenting that the method is a callback.

**Comment by erights** [comment]
From that, I still don't have a precise sense of what you are proposing. Can you explain your proposed semantics of 

```javascript
class Foo {
  handleEvent() => {
      // this is bound to instance
  }
}
```

precisely in terms of an expansion to the existing language? If not, please try to get as close as possible by expansion, and then explain remaining differences in prose.

**Comment by ssube** [comment]
@erights I may have posted in the wrong issue, but I think [this is an expansion](https://github.com/zenparsing/es-function-bind/issues/18#issuecomment-112518867):
```
class Foo {
  bar() => {
    console.log(this instanceof Foo);
  }
}

// becomes

var Foo = (function() {
    var Foo = function() {
        this.bar = function() {
            console.log(this instanceof Foo);
        }.bind(this);
    };

    return Foo;
})();

**Comment by lukescott** [comment]
@erights I copied this from #16,
```javascript
var Foo = (function() {
    var Foo = function() {
        this.handleEvent = this.handleEvent.bind(this);
    };

    Foo.prototype.handleEvent = function() {
        // this will always be instance of Foo
    };

    return Foo;
})();
```
The idea is the same, just different syntax. `method() => {}` instead of `::method() {}`.

**Comment by zenparsing** [comment]
@Naddiseo @lukescott AFAICT, it would mean creating a set of new bound functions for each and every instance (duplicating per-instance state), which is not really desirable.

**Comment by lukescott** [comment]
@zenparsing `::object.func` creates a newly bound function too. By doing `method() => {}` the original method gets replaced with the correct binding. It's also an opt-in feature. You would only do it on event handlers - not every method.

**Comment by Naddiseo** [comment]
@zenparsing, I see what you're saying. Do you know how other languages solve the problem?

**Comment by zenparsing** [comment]
@lukescott But how are users supposed to know when to opt-in, and when not to?   It's likely that, given the choice and imperfect understanding, they will just opt for doing `=>` methods every time.  I think that's what @domenic means by encouraging.

**Comment by sebmck** [comment]
You can already get instance-bound methods with two current class extensions proposals, does there really need to be a third?

[Class properties](https://gist.github.com/jeffmo/054df782c05639da2adb)

```javascript
class Person {
  constructor(firstName, lastName) {
    this.firstName = firstName;
    this.lastName  = lastName;
  }

  getName = () => `${this.firstName} ${this.lastName}`;
}

new Person("John", "Doe").getName.call(null); // "John Doe"
```

[Decorators](https://github.com/wycats/javascript-decorators)

```javascript

```function autobind(target, key, descriptor) {
  return {
    get: function () {
      return descriptor.value.bind(this);
    }
  };
}

class Person {
  constructor(firstName, lastName) {
    this.firstName = firstName;
    this.lastName  = lastName;
  }

  @autobind
  getName() {
    return `${this.firstName} ${this.lastName}`;
  }
}

new Person("John", "Doe").getName.call(null); // "John Doe"

**Comment by erights** [comment]
Yes, this is essentially simulating objects-as-closures, which, if implemented by expansion, requires allocating a function per-method * per-instance. The original class proposals from long ago proposed that support for this be moved into the implementation where it can be optimized. (The E implementation, while low performance in other regards, does demonstrate the needed optimization.)

However, implementors, for understandable reasons, balked. The needed optimization is approximately orthogonal to their existing huge investment in optimizations, and the two would interact leading to even more complex optimization code.

Unlike these earlier class proposals, there's also a semantic problem with the expansion as shown: inheritance. If a subclass overrides handleEvent, the handleEvent in the superclass constructor will simply override the one in the subclass constructor.

**Comment by lukescott** [comment]
@zenparsing The semantics of `() => {}` should be known. So I'm not sure it's true that people will opt for it every time. Also the engine could possibly optimize this by not creating a new function / wrapping the original.

@ssube That's part of the class properties proposal. `method() => {}` seems like a natural expansion of the drafted ES6 spec.

**Comment by ssube** [comment]
@Naddiseo Many languages with this kind of feature make "binding" a function very cheap, as they just push the context onto the stack rather than allocating anything (essentially just always `.apply`ing the function).

@sebmck Encouraging class properties for bound methods works logically (i.e., methods are on the prototype and these bound methods are on the instance) but feels less readable. It's the same function each time, just bound to a different context.

**Comment by erights** [comment]
@sebmck I like the ```@autobind``` decorator a whole lot. Now that I've see it, I think I would actually use it. We could also have it provide a setter that works around the override mistake, much like SES's ```tamperProof``` uses a setter for this purpose.

It has the advantage that it does not prepay a multiplicative allocation cost. It works well with inheritance. If stdized and directly provided by the implementation, where the implementation can recognize and special case it, the implementation could realize that 

```javascript
   person.getName()
```

does not actually require the extra allocation. Skipping that extra allocation would actually be unobservable, which the implementation can know by recognizing its own getter. This does a surprising amount of the optimizations I originally had in mind! Congrats!

However, it still has the same cost that motivated this thread in the first place:

```javascript
    person.getName !== person.getName
```

This cost will not stop me from using this decorator, nor from encouraging others to. I encourage someone to propose it for standardization, so implementations can eventually skip the extra allocation.

Thanks!

**Comment by sebmck** [comment]
@erights The example was a bit concise, but to get `person.getName === person.getName` (and better perf) you could change the `autobind` decorator to:

```javascript
function autobind(target, key, descriptor) {
  return {
    get: function () {
      var fn = descriptor.value.bind(this);
      Object.defineProperty(this, key, { value: fn });
      return fn;
    }
  };
}
```

**Comment by erights** [comment]
Which mutates the instance, which might be frozen. This cure is worse than the disease.

**Comment by sebmck** [comment]
Ah right, good point, hadn't considered that.

**Comment by ssube** [comment]
The instance being frozen shouldn't be an issue if this is done in the constructor (as @lukescott  showed) rather than using a decorator.

**Comment by domenic** [comment]
@Naddiseo yes, that is what I am saying. The fact that React requires such wasteful programming patterns is an unfortunate relic of its architecture, and should not inform language design. (Compare to other frameworks.)

**Comment by erights** [comment]
Doing it in the constructor requires prepaying the multiplicative allocation cost.

**Comment by ssube** [comment]
@erights Is that a huge problem? Allocating objects is (in my experience) already understood to be relatively expensive. It increases the cost, but not in a spot that would be surprising.

**Comment by Naddiseo** [comment]
@sebmck, good point, and I hadn't thought to use a decorator. (slightly off topic) I just did some perfing with the options to see if there were any reasons to choose one over the other. The results are interesting, but I'm not sure how valid they are given that creation could be optimized by the engine.

[Class creation](http://jsperf.com/source-binding-creation): using class properties is a lot faster, docorators are very slow
[Class instantiation](http://jsperf.com/source-binding-instanciation): Decorators fastest, @lukescott's method is slowest
[Bound method calling](http://jsperf.com/bound-calling): class properties are much faster, decoratored are much slower.


@domenic, I accept defeat on the subject. You're right, the framework architecture shouldn't force the language down dark alleys and beat it with a crowbar.

**Comment by erights** [comment]
The multiplicative cost is why the objects-as-closures pattern is not widely enough used, and why early class proposals based on them failed to achieve consensus. Although this decorator would also need optimization, unlike those class proposals, these optimizations should be much more self-contained.

Even without these optimizations, I will be using ```@autobind``` a lot. Thanks!

**Comment by erights** [comment]
Although it is obviously too late, I am trying to remember whether we had considered making the ```@autobind``` behavior the default for classes. I don't remember. If we had, I don't remember why we rejected it.

A shame. Perhaps to reconsider for const classes and for StrongScript classes.

**Comment by erights** [comment]
@ssube @lukescott 
Unlike ```@autobind```, doing it in the constructor does not work with inheritance. This is a fatal flaw that has nothing to do with performance.

**Comment by lukescott** [comment]
@erights @sebmck perhaps I'm missing something, but inheritance doesn't work with autobind either. I get `Person` instead of `Luke`:
```javascript
class Person {
  @autobind
  logthis() {
    console.log(this.constructor.name);
  }
}

class Luke extends Person {
  logthis() {
    super.logthis();
  }
}

function autobind(target, key, descriptor) {
  return {
    get: function () {
      var fn = descriptor.value.bind(target);
      //Object.defineProperty(target, key, { value: fn });
      return fn;
    }
  };
}

var luke = new Luke;
var logthis = luke.logthis;
logthis();
```

Without the autobind and calling the method I get `Luke`:

```javascript
class Person {
  logthis() {
    console.log(this.constructor.name);
  }
}

class Luke extends Person {
  logthis() {
    super.logthis();
  }
}

var luke = new Luke;
luke.logthis();
```

I had to comment out `Object.defineProperty(target, key, { value: fn })` because I was getting `Cannot redefine property: logthis` on babel.

**Comment by zenparsing** [comment]
@lukescott I think you want to bind to `this`, not `target`.  See @sebmck 's implementation.

**Comment by lukescott** [comment]
@zenparsing `this` is undefined. As far as I can tell target is the prototype not the instance.

**Comment by lukescott** [comment]
Take a look at [this REPL](https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&playground=true&code=class%20Person%20%7B%0A%20%20%40autobind%0A%20%20logthis()%20%7B%0A%20%20%20%20console.log(this.constructor.name)%3B%0A%20%20%7D%0A%7D%0A%0Aclass%20Luke%20extends%20Person%20%7B%0A%20%20logthis()%20%7B%0A%20%20%20%20super.logthis()%3B%0A%20%20%7D%0A%7D%0A%0Afunction%20autobind(target%2C%20key%2C%20descriptor)%20%7B%0A%20%20return%20%7B%0A%20%20%20%20get%3A%20function%20()%20%7B%0A%20%20%20%20%20%20var%20fn%20%3D%20descriptor.value.bind(target)%3B%0A%20%20%20%20%20%20%2F%2FObject.defineProperty(target%2C%20key%2C%20%7B%20value%3A%20fn%20%7D)%3B%0A%20%20%20%20%20%20return%20fn%3B%0A%20%20%20%20%7D%0A%20%20%7D%3B%0A%7D%0A%0Avar%20luke%20%3D%20new%20Luke%3B%0Avar%20logthis%20%3D%20luke.logthis%3B%0Alogthis()%3B).

**Comment by lukescott** [comment]
This is the closest thing I can come up with with a getter, and inheritance is still broken:
```javascript
class Person {
  get logthis() {
    let fn = () => {
      console.log(this.constructor.name); 
    };
    Object.defineProperty(this, "logthis",  fn);
    return fn;
  }
}

class Luke extends Person {
  logthis() {
    super.logthis();
  }
}

var luke = new Luke;
var logthis = luke.logthis;
logthis();
```

I get `Object.defineProperty called on non-object`.

My original solution works with inheritance though:

```javascript
var Person = (function () {
  function Person() {
    _classCallCheck(this, Person);
    
    this.logthis = this.logthis.bind(this);
  }

  Person.prototype.logthis = function logthis() {
    console.log(this.constructor.name);
  };

  return Person;
})();

var Luke = (function (_Person) {
  function Luke() {
    _classCallCheck(this, Luke);

    if (_Person != null) {
      _Person.apply(this, arguments);
    }
  }

  _inherits(Luke, _Person);

  Luke.prototype.logthis = function logthis() {
    console.log(this.constructor.name);
    _Person.prototype.logthis.call(this);
  };

  return Luke;
})(Person);

var luke = new Luke();
var logthis = luke.logthis;
logthis();
```

http://jsfiddle.net/8t0ezphd/

I get `Luke` twice in console.log on the original method and the override. The above code suggared is:

```javascript
class Person {
  logthis() => {
    console.log(this.constructor.name);
  }
}

class Luke extends Person {
  logthis() {
    console.log(this.constructor.name);
    super.logthis();
  }
}

var luke = new Luke;
var logthis = luke.logthis;
logthis();
```

Although it is a bit odd that the override doesn't have `=>` as well. But it works without it.

Am I missing something here?

The alternative is going back to `::object.method()` or `object->method()` and using call/apply. But then we have the memorization issue. Imagine if someone used that syntax in a loop without knowing any better.

**Comment by zenparsing** [comment]
@lukescott There may be something going on with babel's translation there.

**Comment by zenparsing** [comment]
@lukescott ah, I think I know what's going on there : )

The overriding function is just a plain value, not a getter.  Need to think though this a bit more...

**Comment by lukescott** [comment]
@zenparsing  If binding is slow, would something like this work better?
```javascript
var Person = (function () {
  function Person() {
    _classCallCheck(this, Person);
    
      var _this = this;
      var _logthis = this.logthis;
      Object.defineProperty(_this, "logthis", {
          configurable: true,
          get: function() {
              var fn = _logthis.bind(_this);
              Object.defineProperty(_this, "logthis", {
                  configurable: true,
                  value: fn 
              });
              return fn;
          }
      });
  }

  Person.prototype.logthis = function logthis() {
    console.log(this.constructor.name);
  };

  return Person;
})();

var Luke = (function (_Person) {
  function Luke() {
    _classCallCheck(this, Luke);

    if (_Person != null) {
      _Person.apply(this, arguments);
    }
  }

  _inherits(Luke, _Person);

  Luke.prototype.logthis = function logthis() {
    console.log(this.constructor.name);
    _Person.prototype.logthis.call(this);
  };

  return Luke;
})(Person);

var luke = new Luke();
var logthis = luke.logthis;
logthis();
```

http://jsfiddle.net/8t0ezphd/1/

**Comment by zenparsing** [comment]
@lukescott OK, so if you autobind with a method with a decorator on a parent class, that doesn't mean that overriding methods are autobound on the child class.  It makes sense now that I've thought about it, but seems a bit surprising at first glance.

**Comment by lukescott** [comment]
@zenparsing Also with the decorator the target is the prototype, not the instance. So that's another issue with the decorator, at least with how it currently works. I think that's how the decorator is supposed to work, right?

**Comment by zenparsing** [comment]
@lukescott It is, and that's the only way it can be really, since no instances exist at the time the decorator is applied.  But for autobind you'd really need to bind to `this`.

**Comment by lukescott** [comment]
Would any of these w/ `method() => {}` be appropriate then?:

http://jsfiddle.net/8t0ezphd/ -- does binding unfront in constructor
http://jsfiddle.net/8t0ezphd/1/ -- applies getter in constructor and does binding on-demand

Inheritance works on both.

**Comment by erights** [comment]
Both have multiplicative allocation costs.

**Comment by lukescott** [comment]
@erights can you elaborate on that? With those examples it only happens in the base class. When you override the method in a subclass the binding still sticks. The second one is on-demand, so it only happens when the method is actually used.

**Comment by erights** [comment]
Are y'all using a plugin or something that I'm missing? These fiddles look like babel output, not es6.


> With those examples it only happens in the base class. 

```javascript
  function Person() {
    _classCallCheck(this, Person);
    
    this.logthis = this.logthis.bind(this);
  }
```

The base class constructor is called per instance, so anything it does per instance is a per instance cost. The ```.bind``` call above allocates. Since you have a line like this for the ```logthis``` method, I presume you would have one for every method. Thus, allocation cost is numInstances*numMethods.

> When you override the method in a subclass the binding still sticks.

I withdraw the objection about inheritance. When the super constructor returns with a bound logthis on the instance, the sub constructor will overwrite it with the bound form of the overriding logthis.  The key is that all the methods themselves are still on the prototype in unbound form, so the overriding bound method can super-delegate to the overridden unbound method on the super prototype.

Because of my previous objects-as-closures based attempts, I overlooked that the per-instance bound methods can still use the unbound methods on the prototype. It is a bit weird that they exist in both places, but it does work.

>  The second one is on-demand, so it only happens when the method is actually used.

```javascript
  function Person() {
    _classCallCheck(this, Person);
    
      var _this = this;
      var _logthis = this.logthis;
      Object.defineProperty(_this, "logthis", {
          configurable: true,
          get: function() {
              var fn = _logthis.bind(_this);
              Object.defineProperty(_this, "logthis", {
                  configurable: true,
                  value: fn 
              });
              return fn;
          }
      });
  }
```

This creates a separate getter per method per instance. It is still just as multiplicative.

**Comment by lukescott** [comment]
@erights I'm using [babel repl](https://babeljs.io/repl/) output and then modifying the output for proof-of-concept.

> The base class constructor is called per instance, so anything it does per instance is a per instance cost. The .bind call above allocates. Since you have a line like this for the logthis method, I presume you would have one for every method. Thus, allocation cost is numInstances*numMethods.

Are you assuming that numMethods is every method in the class? Given that this is an opt-in feature, that number can be 0. For example, the following would be numInstances * 1:

```javascript
class Foo {
  handleEvent() => {
    // this method is always bound to the instance
  }

  someMethod() {
    // this is a normal method. this changes if copied.
  }
}
```

So numMethods should really be numBoundMethods. You only pay the cost for callback methods (event handlers and such). I'm not sure you can get around that cost.

Even with `::object.function` you're paying that cost:
```javascript
element.attachEventListener("change", ::this.handleChangeEvent);
```

And without memorization you pay that cost multiple times in certain situations:

```javascript
for (let element of elements) {
  document.attachEventListener("change", ::this.handleChangeEvent);
}
```

And without memorization this doesn't work:

```javascript
element.removeEventListener("change", ::this.handleChangeEvent);
```

(I'm using the existing prefix `::`, but it could easily be `this->handleChangeEvent` -- makes no difference)

With `method() => {}` you can do just this:
```javascript
element.attachEventListener("change", this.handleChangeEvent);
// and later
element.removeEventListener("change", this.handleChangeEvent);
```

The allocation cost is paid when emulating the feature in pre-ES7. Couldn't `method() => {}` be done differently in the JavaScript engine implementation in such a way where there the original method has a permanent context attached to it? Perhaps something like:

```javascript
Object.defineProperty(Foo.prototype, "logthis", {
  autobind: true,
  value: function() {...}
});
```

**Comment by erights** [comment]
What's the problem again with the original ```@autobind``` annotation exactly as originally defined? In Babel I just tried


```javascript
function autobind(target, key, descriptor) {
  return {
    get: function () {
      return descriptor.value.bind(this);
    }
  };
}

class Person {
  constructor(firstName, lastName) {
    this.firstName = firstName;
    this.lastName  = lastName;
  }

  @autobind
  getName() {
    return `${this.firstName} ${this.lastName}`;
  }
}

class FormalPerson extends Person {
  constructor(title, firstName, lastName) {
    super(firstName, lastName);
    this.title = title;
  }
  
  @autobind
  getName() {
    return `${this.title}. ${super.getName()}`;
  }
}

class SillyPerson extends Person {
  constructor(title, firstName, lastName) {
    super(firstName, lastName);
    this.title = title;
  }
  
  getName() {
    return `${this.title}. ${super.getName()}`;
  }
}

console.log(new FormalPerson("Mr", "John", "Doe").getName.call(null)); // "Mr. John Doe"
console.log(new SillyPerson("Honorable", "John", "Doe").getName()); // "Honorable. Jon Doe"
```

and got the results shown in the comments at the end.

**Comment by erights** [comment]
Let's see if this works:

https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=function%20autobind(target%2C%20key%2C%20descriptor)%20%7B%0A%20%20return%20%7B%0A%20%20%20%20get%3A%20function%20()%20%7B%0A%20%20%20%20%20%20return%20descriptor.value.bind(this)%3B%0A%20%20%20%20%7D%0A%20%20%7D%3B%0A%7D%0A%0Aclass%20Person%20%7B%0A%20%20constructor(firstName%2C%20lastName)%20%7B%0A%20%20%20%20this.firstName%20%3D%20firstName%3B%0A%20%20%20%20this.lastName%20%20%3D%20lastName%3B%0A%20%20%7D%0A%0A%20%20%40autobind%0A%20%20getName()%20%7B%0A%20%20%20%20return%20%60%24%7Bthis.firstName%7D%20%24%7Bthis.lastName%7D%60%3B%0A%20%20%7D%0A%7D%0A%0Aclass%20FormalPerson%20extends%20Person%20%7B%0A%20%20constructor(title%2C%20firstName%2C%20lastName)%20%7B%0A%20%20%20%20super(firstName%2C%20lastName)%3B%0A%20%20%20%20this.title%20%3D%20title%3B%0A%20%20%7D%0A%20%20%0A%20%20%40autobind%0A%20%20getName()%20%7B%0A%20%20%20%20return%20%60%24%7Bthis.title%7D.%20%24%7Bsuper.getName()%7D%60%3B%0A%20%20%7D%0A%7D%0A%0Aclass%20SillyPerson%20extends%20Person%20%7B%0A%20%20constructor(title%2C%20firstName%2C%20lastName)%20%7B%0A%20%20%20%20super(firstName%2C%20lastName)%3B%0A%20%20%20%20this.title%20%3D%20title%3B%0A%20%20%7D%0A%20%20%0A%20%20getName()%20%7B%0A%20%20%20%20return%20%60%24%7Bthis.title%7D.%20%24%7Bsuper.getName()%7D%60%3B%0A%20%20%7D%0A%7D%0A%0Aconsole.log(new%20FormalPerson(%22Mr%22%2C%20%22John%22%2C%20%22Doe%22).getName.call(null))%3B%20%2F%2F%20%22Mr.%20John%20Doe%22%0Aconsole.log(new%20SillyPerson(%22Honorable%22%2C%20%22John%22%2C%20%22Doe%22).getName())%3B%20%2F%2F%20%22Honorable.%20Jon%20Doe%22%0A

**Comment by erights** [comment]
Since
  1.  ```@autobind``` has no multiplicative cost
  1. ```@autobind``` has a semantics I'd prefer just about everywhere compared to the current semantics
  1. Calling a method (```foo.getName()```) is vastly more frequent by both static and dynamic frequency than is extracting a method by name from an instance (```foo.getName```)
  1. With reasonable and local optimizations, ```@autobind``` need not add any costs to method calling
  1. ```@autobind``` adds allocation allocation costs costs only on extracting a method by name, which is rare and does not need to be high performance
  1. ```@autobind```'s semantic cost is only that ```foo.getName !== foo.getName```
  1. ```@autobind``` can also bundle in a workaround for the override mistake,

I will start to use ```@autobind``` essentially all the time. In retrospect, I dearly wish we had made it the default. Thus, I think we should evaluate it under the assumption that it is at least **a** common use case, and perhaps even the more common one.

**Comment by lukescott** [comment]
@erights this doesn't work:
```javascript
var silly = new SillyPerson("Honorable", "John", "Doe");
var getName = silly.getName;
console.log(getName());
```
I get `Cannot read property 'title' of undefined`.

Neither does this:
```javascript
console.log(new SillyPerson("Honorable", "John", "Doe").getName.call(null)); // "Honorable. Jon Doe"
```

This works because you're doing autobind on the subclass:
```javascript
console.log(new FormalPerson("Mr", "John", "Doe").getName.call(null)); // "Mr. John Doe"
```

The autobind needs to be inherited without having to do it on the subclass.

**Comment by erights** [comment]
> The autobind needs to be inherited without having to do it on the subclass.

Why? Even implementation aside, this is not what I would expect. The subclass is overriding it, so the subclass is in control. If the subclass says the method is not auto-bound, I would not expect it to be autobound. IMO, the behavior of your "counter-example" is exactly correct and anything else would be broken.

**Comment by lukescott** [comment]
@erights That is broken behavior. If a method is meant to be a callback for an event handler and someone overrides that method in a subclass the method is still a callback. I would find that behavior very surprising.

For example:
```javascript
class Foo {
  constructor() {
      // this is broken in Foo2!
      element.attachEventListener("change", this.handleChangeEvent);
  }

  @autobind
  handleChangeEvent() {
    // ...
  }
}

class Foo2 extends Foo {
    handleChangeEvent() {
        // ... this is no longer auto bound!
    }
}

new Foo2(document.getElementsByClassName("foo")[0]);
```

I'm not ok with that. Breaking inheritance is broken.

The `::object.method` proposal did not have this problem:
```javascript
class Foo {
  constructor() {
      // still works, still passes a bound function
      element.attachEventListener("change", ::this.handleChangeEvent);
  }

  handleChangeEvent() {
    // ...
  }
}

class Foo2 extends Foo {
    handleChangeEvent() {
        // ...
    }
}

new Foo2(document.getElementsByClassName("foo")[0]);
```

What I suggested is doing is basically what `::object.method` does, it just moves the binding to the class definition and changes `::object.method !== ::object.method` into `object.method === object.method`.

IMO, numInstances * numBoundMethods allocation is acceptable, especially when we're talking about single methods ES5-ES6. I would be more worried about function call performance w/ a bind. ES7 could optimize that allocation away.

**Comment by erights** [comment]
I have run out of time tonight. I hope to get to the rest of it some other time. For now I will respond only to

> That is broken behavior. If a method is meant to be a callback for an event handler 
> and someone overrides that method in a subclass the method is still a callback. I 
> would find that behavior very surprising.

If the subclass overrides it to no longer be a callback for an event handler, then it isn't. If the subclass overrides it to be a dog catcher, then it is a dog catcher. We do not enforce that subclasses must honor the contract associated with the superclass. The "being a callback" aspect of a method is not "final" -- we have no "final".

**Comment by lukescott** [comment]
> If the subclass overrides it to no longer be a callback for an event handler, then it isn't. If the subclass overrides it to be a dog catcher, then it is a dog catcher. We do not enforce that subclasses must honor the contract associated with the superclass. The "being a callback" aspect of a method is not "final" -- we have no "final".

Can you provide a use-case where a callback isn't going to be a callback in a subclass? What % of this use-case makes up the majority of what users are going to want it for? Can you provide justification of forcing someone to re-declare autobind for every subclass?

As per the example in my previous post, knowledge about the method being a callback is in the superclass:

```javascript
class Foo {
  constructor() {
      // this is broken in Foo2!
      element.attachEventListener("change", this.handleChangeEvent);
  }
//...
}
```

What is the likelihood that someone is going to forget to `@autobind` their method in the subclass?

**Comment by erights** [comment]
The number of use cases where a subclass *should* override something from the superclass in a way that violates all expectations associated with the superclass is approximately zero. Along with much of the rest of the world, I believe that the Liskov substitution principle (LSP) is a good design rule that should almost always be followed.

For what % of use cases would you not autobind a method, once autobinding is available and cheap. I think this is also approximately zero. Autobinding should have been the default. Since we missed this window for plain classes, our options are
  1. Add a checking rule to jshint that any method on a plain class not annotated with ```@noautobind``` be annotated with ```@autobind```.
  1. Introduce other non-plain classes, like const classes or StrongScript classes, where we can get this default right.
  1. Invent a class annotation with similar effect: on such an annotated class, all methods not annotated with ```@noautobind``` default to autobound.

All of these have the effect of making a silent non-autobinding by mistake into a detectable hazard. This should include accidentally forgetting to autobind an override of an autobound inherited method.

When would you ever use ```@noautobind``` on purpose? I am sure there are cases, but I currently can't think of any.

> Can you provide justification of forcing someone to re-declare autobind for every subclass?
> What is the likelihood that someone is going to forget to @autobind their method in the subclass?

Can you provide justification for forcing someone to say "const", "let", "var", or "function" on every variable declaration?
What is the likelihood that someone will forget to say "const", "let", "var", or "function" when declaring a variable, but instead inadvertently write a plain assignment? I know I have done this more times than I can count. I am very happy that strict mode and now modules have made this hazard into a reported error.

**Comment by lukescott** [comment]
Class methods should have been bound by default: In principle, I agree. The reason this was not done by default is because it is not cheap, especially when emulating in ES5.

> 1. Add a checking rule to jshint that any method on a plain class not annotated with @noautobind be annotated with @autobind.

That's crazy. Most methods are not passed as callbacks. Binding every method would double the call time of for _all_ methods.

An event handler is a very specific use case where you _have to bind_. A use case we're actually trying to solve. I don't think it's unreasonable to say that any method defined as bound should stay bound in a subclass that chooses to override that method.

> 2. Introduce other non-plain classes, like const classes or StrongScript classes, where we can get this default right.

And what does that look like with `class Foo {}`?

> Can you provide justification for forcing someone to say "const", "let", "var", or "function" on every variable declaration? What is the likelihood that someone will forget to say "const", "let", "var", or "function" when declaring a variable, but instead inadvertently write a plain assignment? I know I have done this more times than I can count. I am very happy that strict mode and now modules have made this hazard into a reported error.

Apples and oranges. A variable context is in the same function - easier to spot. A superclass implementation is in another file or library. Overriding a superclass method should not break the method.

**Comment by erights** [comment]
> The reason this was not done by default is because it is not cheap, especially when emulating in ES5.

I was there. I don't remember whether we considered it and if we did, I don't remember why we rejected it. It is indeed possible that it was a performance concern that I'm not getting.

However, I am confident that the performance of a transpiled-to-ES5 implementation was not a reason. We do not make decisions on that basis.

> That's crazy. Most methods are not passed as callbacks. Binding every method 
> would double the call time of for all methods.

It is crazy if it actually doubles the call times for methods. I bet that implementations, by recognizing their own getter, can keep the call times to essentially what they are now. In fact, implementations should eventually treat the bound case as the expected case, and if necessary make the unbound case pay for being the odd duck.

> I don't think it's unreasonable to say that any method defined as bound should stay
> bound in a subclass that chooses to override that method.

Let's say we had gotten the default right and you had to say ```@noautobind``` to get the current behavior. In that case, if a superclass writes an unmarked (and therefore autobound) method and the subclass overrides it with an ```@noautobind```, should the subclass' annotation be ignored since the superclass said (by silence) that the method should be autobound?

> And what does that look like with class Foo {}?

```const class Foo {}```

> Apples and oranges. A variable context is in the same function - easier to spot. 
> A superclass implementation is in another file or library. Overriding a superclass 
> method should not break the method.

Now that we have the option, choosing not to autobind a method is bad practice and should generally be assumed to break things. Once it is cheap, why would you ever do it? This is the central question that is so far unanswered.

**Comment by BrendanEich** [comment]
I hate OOP. It's all Java's fault that JS isn't this- and new-free, first-class functions only :-P.

Going down the OOP rabbit holes (plural), I see why @lukescott wants @autobind to be inherited even when overriding. Or really, everyone wants StrongScript classes. ES4 classes (AS3 in Flash/Flex), IIRC, worked the desired way.

@erights, we didn't add auto-binding to ES6 classes because we did not want to impose a hidden cost not found in prototypal-pattern (desugared) classes in ES5.

If a new mode a la strong/sound can help, great. Otherwise I think @erights and @zenparsing are on surer footing arguing for less OOP happiness as the price of straightforward (as if by desugaring) if not pellucid semantics: @autobind not inherited, and obj.foo !== obj.foo.

/be

**Comment by erights** [comment]
@BrendanEich This sounds plausible, but do you remember if we ever explicitly considered it?

> pellucid

I had to look it up. Nice word. Thanks ;).

**Comment by erights** [comment]
@BrendanEich To clarify, I know we explicitly considered and rejected doing it automagically in the implementation, in such a way that ```obj.foo === obj.foo```. This was part of my early class proposals based on making objects-as-closures or traits efficient. This is also closer to what I gather was in ES4. This would indeed have required new orthogonal implementation machinery that implementers were right to balk at. I understand why it died.

However, the getter technique of ```@autobind```, done at the price of ```obj.foo !== obj.foo```, is the idea that strikes me as new. It seems free of the problems that killed everything else I remember, and it does it within the ES5 object model with no new semantic state. Do you remember if we had ever explicitly considered it?

**Comment by BrendanEich** [comment]
We discussed it over the many years, at least a couple of times. I don't have time to search meeting notes on esdiscuss.org right now :-P.

/be

**Comment by erights** [comment]
Not asking for a record, just your memory. Thanks!

**Comment by BrendanEich** [comment]
The getter idea wasn't framed via decorators, but some of us have been around this corner of the design space and (I recall) we still didn't like the implicit cost. Also not observed in the wild in class desugarings or idiomatic prototypal pattern instances.

Decorators help. Opt-in FTW.

**Comment by lukescott** [comment]
@erights 

> Now that we have the option, choosing not to autobind a method is bad practice and should generally be assumed to break things. Once autobinding is cheap, why would you ever @noautobind? This is the central question that is so far unanswered.

I agree. You would never want `@noautobind`. If it was cheap, could you change classes in ES7 to `@autobind` by default? How much code would break? Does anyone rely on `@noautobind`?

This benchmark is not very encouraging though: http://jsperf.com/bind-vs-emulate/13.

```
Chrome 43.0.2357.124 on OS X 10.10.3
bind - 5,594,117 - 99% slower
emulate - 92,910,404 - 79% slower
simple - 93,837,139 - 79% slower
self - 441,522,153
```

At least under the assumption that every method was `@autobind` by default and emulated in ES5 (ES6?). That penalty is easier to swallow with opt-in on a per-method basis.

Re: `@autobind` on both superclass and subclass methods, can you take a [look at this issue](https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&playground=true&code=function%20autobind(target%2C%20key%2C%20descriptor)%20%7B%0A%20%20return%20%7B%0A%20%20%20%20get%3A%20function%20()%20%7B%0A%20%20%20%20%20%20console.log(%22get%20bind%22)%3B%0A%20%20%20%20%20%20let%20fn%20%3D%20descriptor.value.bind(this)%3B%0A%20%20%20%20%20%20Object.defineProperty(this%2C%20key%2C%20%7B%0A%20%20%20%20%20%20%20%20configurable%3A%20true%2C%0A%20%20%20%20%20%20%20%20value%3A%20fn%0A%20%20%20%20%20%20%7D)%3B%0A%20%20%20%20%20%20return%20fn%3B%0A%20%20%20%20%7D%0A%20%20%7D%3B%0A%7D%0A%0Aclass%20Person%20%7B%0A%20%20constructor(firstName%2C%20lastName)%20%7B%0A%20%20%20%20this.firstName%20%3D%20firstName%3B%0A%20%20%20%20this.lastName%20%20%3D%20lastName%3B%0A%20%20%7D%0A%0A%20%20%40autobind%0A%20%20getName()%20%7B%0A%20%20%20%20return%20%60%24%7Bthis.firstName%7D%20%24%7Bthis.lastName%7D%60%3B%0A%20%20%7D%0A%7D%0A%0Aclass%20FormalPerson%20extends%20Person%20%7B%0A%20%20constructor(title%2C%20firstName%2C%20lastName)%20%7B%0A%20%20%20%20super(firstName%2C%20lastName)%3B%0A%20%20%20%20this.title%20%3D%20title%3B%0A%20%20%7D%0A%20%20%0A%20%20%40autobind%0A%20%20getName()%20%7B%0A%20%20%20%20return%20%60%24%7Bthis.title%7D.%20%24%7Bsuper.getName()%7D%60%3B%0A%20%20%7D%0A%7D%0A%0Avar%20person%20%3D%20new%20FormalPerson(%22Mr%22%2C%20%22John%22%2C%20%22Doe%22)%3B%0A%0Aconsole.log(person.getName.call(null))%3B%20%2F%2F%20Mr.%20John%20Doe%0Aconsole.log(person.getName.call(null))%3B%20%2F%2F%20John%20Doe.%0A)? I tried to memorize `.bind` as @sebmck suggested, but the method ends up being overwritten by the super-class implementation. Regardless, you end up with a `.bind` on each implementation ("get bind" logs twice). If doing `.bind` in the constructor, it only happens once and the correct method is always used.

> Brendan Eich: Or really, everyone wants StrongScript classes.

Here here!

**Comment by erights** [comment]
Changing plain classes to autobind by default will never happen. I never suggested that. If you think I did, please reread my suggestions.

**Comment by erights** [comment]
Re [look at this issue](https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&playground=true&code=function%20autobind(target%2C%20key%2C%20descriptor)%20%7B%0A%20%20return%20%7B%0A%20%20%20%20get%3A%20function%20()%20%7B%0A%20%20%20%20%20%20console.log(%22get%20bind%22)%3B%0A%20%20%20%20%20%20let%20fn%20%3D%20descriptor.value.bind(this)%3B%0A%20%20%20%20%20%20Object.defineProperty(this%2C%20key%2C%20%7B%0A%20%20%20%20%20%20%20%20configurable%3A%20true%2C%0A%20%20%20%20%20%20%20%20value%3A%20fn%0A%20%20%20%20%20%20%7D)%3B%0A%20%20%20%20%20%20return%20fn%3B%0A%20%20%20%20%7D%0A%20%20%7D%3B%0A%7D%0A%0Aclass%20Person%20%7B%0A%20%20constructor(firstName%2C%20lastName)%20%7B%0A%20%20%20%20this.firstName%20%3D%20firstName%3B%0A%20%20%20%20this.lastName%20%20%3D%20lastName%3B%0A%20%20%7D%0A%0A%20%20%40autobind%0A%20%20getName()%20%7B%0A%20%20%20%20return%20%60%24%7Bthis.firstName%7D%20%24%7Bthis.lastName%7D%60%3B%0A%20%20%7D%0A%7D%0A%0Aclass%20FormalPerson%20extends%20Person%20%7B%0A%20%20constructor(title%2C%20firstName%2C%20lastName)%20%7B%0A%20%20%20%20super(firstName%2C%20lastName)%3B%0A%20%20%20%20this.title%20%3D%20title%3B%0A%20%20%7D%0A%20%20%0A%20%20%40autobind%0A%20%20getName()%20%7B%0A%20%20%20%20return%20%60%24%7Bthis.title%7D.%20%24%7Bsuper.getName()%7D%60%3B%0A%20%20%7D%0A%7D%0A%0Avar%20person%20%3D%20new%20FormalPerson(%22Mr%22%2C%20%22John%22%2C%20%22Doe%22)%3B%0A%0Aconsole.log(person.getName.call(null))%3B%20%2F%2F%20Mr.%20John%20Doe%0Aconsole.log(person.getName.call(null))%3B%20%2F%2F%20John%20Doe.%0A), as already discussed, that code does not work on frozen instances.

**Comment by erights** [comment]
> If doing .bind in the constructor, it only happens once and the correct method is always used.

If doing .bind in the constructor, it still happens twice, but the subclass constructor overwrites the binding from the superclass constructor so the correct method is indeed always used. This was the "does not work with inheritance" objection that I already withdrew. You were right about that.

**Comment by lukescott** [comment]
I did not. I was just asking the question: Who relies on `@noautobind` behavior? I know that making a change like that would be a breaking change. But, if as you say, such use cases are approximately zero, how much harm would such a change be? That is, writing every class as `strict class Foo {}` vs fixing `class Foo {}`.

> Re look at this issue, as already discussed, that code does not work on frozen instances.

Ok, so without memorization on `@autobind` the cost per call to getName is:
- n getter calls
- n bind calls
- n calls to bound methods
n = number of times `@autobind` is used on the same method.

> If doing .bind in the constructor, it still happens twice, but the subclass constructor overwrites the binding from the superclass constructor so the correct method is indeed always used. This was the "does not work with inheritance" objection that I already withdrew. You were right about that.

Just want to make sure I'm on the same page: Are you saying that `.bind` happens twice in the constructor w/ opt-in? With the example I provided it only happens once:

http://jsfiddle.net/8t0ezphd/2/

Assuming that the super-class does the binding and the subclass does not.

**Comment by erights** [comment]
> how much harm would such a change be?

I don't know. It is a question worth asking. Data would be great. TC39 loves data, and data has a huge influence on our decisions.

In the absence of data, we rely on our intuitions informed by experience. We do occasionally make a bet, even without data, that a technically breaking change is not actually breaking, and that it is worth the cost of trying and seeing what happens. We have been very successful at that -- we have gotten away with many technically breaking changes without actually breaking the web. Of those we've decided to try, we have had to backtrack very few.

However, looking at the bets that we have and have not decided to try without data, I am rather sure this one is under threshold. The payoff is actually rather high, which is great, but my intuition about the probability of breakage is that it is also too high for us to take a chance on. Again, data could completely change the situation for the better.

> n getter calls

What is n?

> Just want to make sure I'm on the same page ... Assuming that the super-class 
> does the binding and the subclass does not.

Sorry, I was responding to the other case -- the one where I consider your proposal to be doing the right thing, where both the subclass and superclass say they want autobinding. You are correct that for your only-super-autobinds case the binding only happens once and it still binds the right method.

However, I consider this a bug, not a feature. Looking at the subclass by itself, let's say we see a class whose getName method does not invoke super.getName. By normal understanding its getName behavior should then be self contained -- I should be able to figure out what it is locally. I look for other delegations from sub to super. Of course I necessarily see one in the constructor. I look at the super constructor. I see no mention of getName. Thus, it is surprising to me that the superclass has still intervened in the meaning of my getName.

For both of your cases, ```Object.getOwnPropertyNames(foo)``` has extra members. This is weird. It is at least as weird as ```foo.getName !== foo.getName```. Since each proposal has such a weirdness, we can consider these a draw on this criteria.

**Comment by lukescott** [comment]
> I don't know. It is a question worth asking. Data would be great. TC39 loves data, and data has a huge influence on our decisions.
> ...
> However, looking at the bets that we have and have not decided to try without data, I am rather sure this one is under threshold. The payoff is actually rather high, which is great, but my intuition about the probability of breakage is that it is also too high for us to take a chance on. Again, data could completely change the situation for the better.

How would we collect data on this?

> What is n?

Number of levels `@autobind` is applied. So in the case of getName(), this would be 2. In the code sample I provided earlier "get bind" was emitted 2 times per call. Could be 3 if FormalPerson was subclassed and getName() was again reimplemented w/ autobind().

> However, I consider this a bug, not a feature. Looking at the subclass by itself, let's say we see a class whose getName method does not invoke super.getName. By normal understanding its getName behavior should then be self contained -- I should be able to figure out what it is locally.

I suppose this is true for the getName() case. But getName() is also not a use-case where `@autobind` would be useful. No one will notice whether or not the method is bound or not. The method will likely be called off the class instance. This will be true for a majority of class methods.

> I look for other delegations from sub to super. Of course I necessarily see one in the constructor. I look at the super constructor. I see no mention of getName. Thus, it is surprising to me that the superclass has still intervened in the meaning of my getName.

An inherited callback handler, such as handleChangeEvent(), would not be surprising. 

The more we discuss this the more of an odd duck any form of `@autobind` is to me, inherited or not. It's surprising to me that any method would lose binding to the instance, whether it was an inherited method with `@autobind` or a brand new method.

I would rather pursue fixing class methods. We could test that theory with a babel/traceur plugin/option.

**Comment by erights** [comment]
> How would we collect data on this?
> I would rather pursue fixing class methods. We could test that theory 
> with a babel/traceur plugin/option.

Excellent. That seems like a great way to start.

**Comment by ssube** [comment]
@lukescott 
> Who relies on @noautobind behavior? I know that making a change like that would be a breaking change. But, if as you say, such use cases are approximately zero, how much harm would such a change be?

If I understand this correctly (bind methods to their instance by default, decorate to keep variable `this`?), a lot of older DOM-related libraries like jQuery use that behavior as the default. One of the examples for [`on`](http://api.jquery.com/on/) is:
```
$( "#dataTable tbody tr" ).on( "click", function() {
  console.log( $( this ).text() );
});
```

How often that crops up in as a class method (formal ES6 class or IIFE pseudo-class), I'm not sure.

**Comment by lukescott** [comment]
@ssube 

This would only be for ES6 classes. JavaScript objects would not change.

> If I understand this correctly (bind methods to their instance by default, decorate to keep variable this?), a lot of older DOM-related libraries like jQuery use that behavior as the default. One of the examples for on is:

jQuery wraps DOM elements. A jQuery wrapper is returned from `$( "#dataTable tbody tr" )` containing 0 or more elements. `on` is being called from the wrapper, so it is bound to the wrapper.

I'm not sure if this works:
```javascript
var result = $( "#dataTable tbody tr" );
result.on( "click", function() {
  console.log( $( this ).text() );
});
```

But if it does they are likely doing their own binding. Regardless, they aren't using ES6 classes anyway.

The method binding is related to the `::object.method` part of the original proposal.

The `object::function` part concerns jQuery. In #18 I proposed something like this:
```javascript
import {on, text} from "jquery";

document.querySelectorAll("#dataTable tbody tr").(on, "click", () => {
  console.log(this.(text));
});
```

That would eliminate jQuery's need for the wrapper.

**Comment by ssube** [comment]
@lukescott I wasn't worried about jQuery wrapping elements, but folks passing class methods to jQuery events. For example:
```
class Foo {
  bar() {
    console.log(this);
  }

  attach(jq) {
    jq.on('click', this.bar);
  }
}
```
Today, that will log the element that triggered the event rather than the surrounding class (see [this fiddle](http://jsfiddle.net/hsg5yg6j/)).

With `@noautobind` as the decorator and autobinding as the default, that behavior will change, no? Or am I misunderstanding the idea?

**Comment by Naddiseo** [comment]
@ssube, IIRC that happens without jquery as well, normal `addEventListener` will bind the element as `this`, so auto-binding class methods would also change that behaviour. However, as @erights mentioned, getting evidence as to whether that would be a problem is probably what needs to happen before making a decision one way or another.

**Comment by lukescott** [comment]
@ssube Gotcha. Yep, that is correct. Although I would find one of these to be more likely:
```javascript
class Foo {
  bar = (e) => {
    console.log(e.toElement);
  }

  attach(jq) {
    jq.on('click', this.bar);
  }
}
```
```javascript
class Foo {
  constructor() {
    this.bar = this.bar.bind(this);
  }

  bar(e) {
    console.log(e.toElement);
  }

  attach(jq) {
    jq.on('click', this.bar);
  }
}
```

As soon as you need the instance you have to do something like the above.

At this stage we're only talking about experimenting with the idea of making the change.

**Comment by BrendanEich** [comment]
For future reference:

https://esdiscuss.org/topic/methods-automatic-binding-on-read (3 years old)

https://github.com/gbracha/generalizedTearOffs/blob/master/proposal.md (fresh Dart work)

/be

**Comment by WebReflection** [comment]
I've spot this conversation only recently via ES mailing list, I wonder why a simple mixin-like approach hasn't been discussed yet: it solves inheritance gotcha, it could work without extra costs (cost is at runtime only if the utility is accesed)

```js
// the mixin
var boundable = (function (wm) {
  return function boundable(Class) {
    Object.defineProperty(
      Class.prototype,
      'boundTo',
      {
        writable: true,
        configurable: true,
        value: function boundTo(method) {
          var
            callback = typeof method === 'function' ?
              method : this[method],
            m = wm.get(this) || (wm.set(this, m = {
              original: [],
              bound: []
            }), m),
            i = m.original.indexOf(callback)
          ;
          if (i < 0) {
            i = m.original.push(callback) - 1;
            m.bound[i] = callback.bind(this);
          }
          return m.bound[i];
        }
      }
    );
  };
}(new WeakMap));

// the test
@boundable
class Test {
  constructor(name) {
    this.name = name;
  }
  getName() {
    return this.name;
  }
}

var me =  new Test('Andrea');
console.log(
  me.boundTo('getName') ===
  me.boundTo(me.getName)
); // true

var getName = me.boundTo('getName');
console.log(getName()); // Andrea
```

Above method has been successfully used for long time  in [eddy](https://github.com/WebReflection/eddy#objectboundtomethod) through the global `Object.prototype` but since  these decorators are  apparently a thing I think the  suggested snippet might be a solution to the very common issue?

Otherwise I think I would vote +1 for the `::obj.method` to return the same bound method each time, which is the least surprise in  my opinion and  yes,  many developers keep  loosing tracks of their listeners with arrow function and bind since about ever :-(

Just my little contribution,  if any.
Best Regards

**Comment by jussi-kalliokoski** [comment]
I made [weakbind](https://github.com/jussi-kalliokoski/weakbind) as a proof of concept for this (and to be possibly used by the babel transform for this syntax) and if you look at the [benchmarks](https://github.com/jussi-kalliokoski/weakbind#benchmark), it looks very promising - and the benchmarks are much simpler, especially the React one in terms of render tree, than real world applications, so the benefit is likely to be even greater in the real world. It uses a two-dimensional WeakMap which allows it to nicely provide idempotency without causing memory leaks or additional bookkeeping.

**Comment by zenparsing** [comment]
@jussi-kalliokoski Cool.  As discussed upthread, for the syntactic use case we'd want the memoized bound functions to be frozen (in order to prevent communication channels).

**Comment by WebReflection** [comment]
Since this has been brought up again in es-discuss ML, I'd like to ask @erights and others a few questions:

  1. I've never ever needed two different version of a bound function/method for the same object. The only reason that happened is because that's how `.bind` was specified, and it's the No. 1 footgun when  it comes  to assign at runtime a method to a generic event ( without using `hadnleEvent` that indeed doesn't need that, but that's just one story,  not the whole scenario,  and nodejs has no such thing ). Accordingly, does anyone in here have a valid use case for having two different versions of the same bound method to the same object is desired?
  2. using `::obj.fn` just to shortcut `fn.bind(obj)` seems to me a huge missed opportunity. The amount of saved chars and the explicit intent and well known behavior is not worth trapping forever `::` as shortcut that won't fix concretely any problem. Are we sure it's actually a good idea to move on this way instead of bringing in something new,  something better, something needed way more than just bind?
  3. the security implication thing looks like a bit overlooked. Every method, inherited or not, of every object has the same problem. Why would we block  something developers  keep expecting and asking for the `::` syntax? If we can access the object any other form of communication channel would be possible and developers these days use Symbols for these things thinking these are somehow private or protected. Can we please be realistic on what `::` would like to solve as real need for developers, and avoid pushing back anything that is not fully frozen down  to its inherited properties? If not, why isn't a new kind of reflection method such `Reflect.petrify(obj)` considered or proposed, where the object gets promoted to a Proxy version of itself that won't accept any change to the object itself and any of its inherited properties and methods? I feel like `Object.freeze` went out with a huge handicap nobody wants to fix, and the side effect is that we cannot think forward about utilities truly needed and wanted, things like `::obj.method === ::obj.method` that would make developers life easier and free from wrong expectations. Then they'll  have every other way they want to footgun themselves.

Thanks foor any sort of outcome,  reply, consideration, and/or enlightenment.

Best Regards

**Comment by isiahmeadows** [comment]
I would actually be inclined to disagree with the idea `::obj.foo` should be identical to `::obj.foo`. Although I see potential footgun material, what good is it supposed to do? Not to mention it would require significant boilerplate to polyfill, and a *lot* of reference counting. It would hardly improve speed, as it is already very cheap and easy to create a fast path in engines for. Allocate a closure with two pointers (host and function), and set the `[[Call]]` and `[[Construct]]` behavior to directly call the methods with little memory or speed overhead.

As for memoization, it's not much of a speed or space optimization:

1. It would actually increase space, because not only the typical reference counting, you now have an unordered weak map for any object with a bound method. This would use more pointers than any constructed closure.
2. It would be slower to create and recall. You have to include extra logic to ensure that the new closure isn't recreated. And testing for existence in a weak collection isn't as quick as a simple key-value map.
3. Because the common case is throwaway methods, it would end up having to construct new instances more often than not. As a weak collection, the throwaway method would be recycled early in most GCs. And adding/removing an instance from a weak collection is slower than constructing a bound method with a few smart pointers.

(Yes, I have a clue how engines implement JS. I've read most of the V8 code base, and large sections of SpiderMonkey. I also have a clue how several other languages' runtimes are implemented.)

And now for the other reason: if it's memoized, then it doesn't maintain the usual invariant that functions are never identical. That will become another footgun, although less common. And also, where `::obj.method` is currently equivalent to `obj.method.bind(obj)`, would one expect that `obj.method.bind(obj) === obj.method.bind(obj)`?

---

And as for discussion on a related proposal, I will kindly refer you to #26 (which is still somewhat active).

**Comment by WebReflection** [comment]
> I see potential footgun material
potential VS common use cases

> because the common case is throwaway methods
that's a side effect of what developers think bind does, not their intention in most of the case.

Developers do believe if they add `obj.method.bind(obj)` as listener, as example, they can remove it via `obj.method.bind(obj)` which is the number 1 footgun about bind.

Moreover, please tell me a single use case for having `obj.method.bind(obj)` being different from `obj.method.bind(obj)` ... because AFAIK nobody needs that and most developers think already that's how it is, forgetting to store references, unable to "regret" and update their state.

Last, but not least, `::` is fresh new syntax, the same it was for `=>` fat arrow ... there was nothing like fat arrow before, that doesn't mean  fat arrow is bad or anything, that simply means fat arrow is born more mature and thoughtful ... and so should be `::obj.foo` proposal.

**Comment by isiahmeadows** [comment]
I'm more speaking from implementation standpoint. And I see massive memory
leak potential with that semantic. That's my main concern.

**Comment by zenparsing** [comment]
I think @WebReflection 's argument regarding `addEventListener/removeEventListener` is convincing, however as we've explored above securely memoizing the extracted method seems like it will be rather heavy-weight.

Also, it creates a divergence from the current behavior of "bind"; it means that this proposal becomes something more than just syntactic sugar.  There's certainly nothing wrong with adding core capabilities to the language, but the current proposal just "slides right in" to the existing object model.  I find that to be really attractive.

Additionally, it would create a semantic divergence between the unary form and the binary form of the operator, which we would prefer to avoid.

In response to the `addEventListener/removeEventListener` issue, in my opinion I would like JS/HTML to continue moving toward better async APIs which don't require callback registration/de-registration patterns.  For consuming UI event streams, I like [Observables](https://github.com/zenparsing/es-observable).

**Comment by isiahmeadows** [comment]
@WebReflection @zenparsing That's a good point as well. My big worry about memory leaks are this:

```js
// obj.js
import {EventEmitter} from "events"
export const obj = {a() { return this }}
export const ee = new EventEmitter()
```

```js
// foo.js
import assert from "assert"
import {ee, obj} from "./obj.js"
let foo = ::obj.a
ee.on("ev", function listener(f) {
  assert.strictEqual(f, foo)
  foo = null
  ee.removeListener("ev", listener)
})
```

```js
// bar.js
import {ee, obj} from "./obj.js"
ee.emit("ev", ::obj.a)
```

```js
// main.js
import "./foo.js"
import "./bar.js"
```

My question: does a reference to `::obj.a` exist in main.js after the imports? Logically, it shouldn't, but this is actually a very complex problem to algorithmically solve, and in fact, this could result in memory leaks if the GC fails to realize `::obj.a` is no longer accessible.

**Comment by zenparsing** [comment]
I think we've exhausted this thread, thanks everyone.  If anyone wants to continue discussing any of these points, feel free to open a new issue.
